packages/vite/src/node/plugins/asset.ts

```diff
@@ -222,10 +222,32 @@ async function fileToBuiltUrl(
     if (!map.has(contentHash)) {
       const basename = path.basename(file)
       const ext = path.extname(basename)
-      const fileName = path.posix.join(
+      let fileName = path.posix.join(
         config.build.assetsDir,
         `${basename.slice(0, -ext.length)}.${contentHash}${ext}`
       )
+
+      const { output } = config.build?.rollupOptions
+      // Only the object format is currently considered here.
+      if (output && !Array.isArray(output)) {
+        const assetFileNames = output.assetFileNames
+        // e.g assetFileNames: dir/[name].[ext]
+        if (typeof assetFileNames === 'string') {
+          fileName = assetFileNames
+            .replace(/\[name\]/, basename.slice(0, -ext.length))
+            .replace(/\[ext\]/, ext.replace(/^\./, ''))
```

Comment:
**question:** how does this behave with something that has two dots in filename?
Like `dir/file.d.ts` or `dir/file.first.sec` or even hidden files (if this is a thing 🤷) `.file`

Reply:
I don't think it affects it here. This is also the logic of handling assets before. Except for .file files, the beginning is not supported, and vite itself does not support assets files beginning with `.`